text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1019 @@
+
+Many of the features from these discussions which are core to Glimmer
+components have already landed in Ember, including: Angle-bracket invocation,
+named arguments, element modifiers, and component managers.
+
+Glimmer components are the final piece of that puzzle. They include the last
+of the major features that were discussed during the original Angle Brackets
+RFC, and holistically those features make a much simpler and more ergonomic
+component API. Taken alone, they are an incremental change. Their individual
+features aren't _that_ much more than what we currently have in Ember today. But
+as a whole they represent the culmination of multiple years of design work and
+discussion by the Ember community, and the collective attention to detail and
+care of all of our community members.
+
+## Terminology
+
+* The **Glimmer VM** is the underlying rendering engine which is used by
+  _Ember.js_ and _Glimmer.js_.
+* **Glimmer.js** is a thin wrapper on top of the _Glimmer VM_ which exposes a
+  much simpler API compared to Ember. Historically it has been used to prove out
+  ideas and implementations before bringing them into Ember via RFC, and has
+  been used to write applications which don't require the full feature set of
+  Ember.
+* **Glimmer components** are a newly proposed component API which draw from the
+  experimental APIs provided in Glimmer.js, and Ember.js via
+  [sparkles-component](https://github.com/rwjblue/sparkles-component).
+* **Classic components** refer to the standard component API at the time of this
+  RFC, which have been available in Ember in some form since v1.
+* **Tracked properties** refer to a new method of change tracking which is being
+  proposed in a separate RFC, parallel to this one.
+
+## Motivation
+
+Glimmer components are a simpler base component class that encourage template
+oriented design and thinner class definitions. They are easier to understand in
+isolation, and require less knowledge of the framework to use effectively.
+
+This example shows a component written with the classic component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+{{#if (eq this.type 'image'}}
+  <img src={{@post.imageUrl}} title={{@post.imageTitle}}/>
+{{/if}}
+
+{{@post.text}}
+```
+```js
+// components/post.js
+export default Component.extend({
+  tagName: 'section',
+  classNames: ['post'],
+  classNameBindings: ['type'],
+  ariaRole: 'region',
+
+  /* Arguments */
+  post: null,
+
+  type: readOnly('post.type'),
+
+  didInsertElement() {
+    if (this.type === 'image') {
+      setupImageOverlay(this.element.querySelector('img'));
+    }
+  }
+});
+```
+
+And here is an equivalent component written with the Glimmer component API:
+
+```hbs
+<!-- templates/components/post.hbs -->
+<section ...attributes role="region" type={{@post.type}} class="post {{@post.type}}">
+  {{#if (eq @post.type 'image')}}
+    <img {{did-render this.didInsertImage}} src={{@post.imageUrl}} title={{@post.imageTitle}} />
```

Comment:
I'm wondering if `{{did-render}}` should be called something different. If I remember correctly the timing of this would be "installed on element but not painted". I also don't know if the state of the element is "in-dom". Might be worth checking.

Reply:
@chadhietala 
https://github.com/emberjs/ember.js/blob/88319874d3b8ec2eab92e99de3fc6b48feddf26c/packages/%40ember/-internals/glimmer/tests/integration/components/life-cycle-test.js#L231 (thank you for writing these awesome tests 😃 ), `didRender` is `inDOM` (fully post render).

Reply:
definitely open to naming suggestions here, it's also different enough from the `didRender` lifecycle hook on classic components that we may want to distinguish it